### PR TITLE
feat: added retry and sleep methods

### PR DESCRIPTION
**Pull request by verbanicm**

### CODE REVIEW

File: src/retry.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  backoff?: number;
+}
+
+/**
+ * withRetry a simple implementation of a retry mechanism with exponential backoff
+ *
+ * @param fn. A function to retry on failure
+ * @param opts. The retry options
+ */
+export async function withRetry(fn: () => Promise<unknown>, opts?: RetryOptions): Promise<unknown> {
```
Comment (sethvargo) [inline src/retry.ts:33]: Can you type this?

```typescript
export async function withRetry<T>(fn: () => Promise<T>, opts?: RetryOptions): Promise<T> {
```

File: src/retry.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  const DEFAULT_ATTEMPTS = 3;
+  const DEFAULT_BACKOFF = 30 * 1000; // 30 seconds
```
Comment (sethvargo) [inline src/retry.ts:35]:
```suggestion
  const DEFAULT_BACKOFF_MILLISECONDS = 30 * 1000; // 30 seconds
```

File: src/retry.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  let callCount = 0;
+  while (callCount < retryOptions.attempts) {
+    try {
+      return await fn.call(null);
+    } catch (e) {
+      ++callCount;
+      if (callCount <= retryOptions.attempts) {
+        await sleep(retryOptions.backoff);
+        retryOptions.backoff = retryOptions.backoff * 2;
```
Comment (sethvargo) [inline src/retry.ts:51]: You probably want to do a fibonacci backoff instead of exponential here.

File: src/time.ts [hunk 4]
```diff
@@ -76,3 +76,16 @@ export function parseDuration(input: string): number {
...
+/**
+ * sleep waits for a specified duration in milliseconds
+ *
+ * @param ms. Duration in milliseconds to sleep.
+ */
+export async function sleep(ms = 0): Promise<void> {
+  return new Promise((resolve) => {
+    setTimeout(() => {
+      resolve();
+    }, Number(ms));
```
Comment (sethvargo) [inline src/time.ts:89]: Should the type be explicit on the function instead?

File: src/time.ts [hunk 5]
```diff
@@ -76,3 +76,16 @@ export function parseDuration(input: string): number {
 
   return total;
 }
+
+/**
+ * sleep waits for a specified duration in milliseconds
```
Comment (sethvargo) [inline src/time.ts:81]:
```suggestion
 * sleep waits for a specified duration in milliseconds as a promise.
```

File: src/time.ts (hunk 4 above)
Comment (verbanicm) [reply src/time.ts:89]: Linter threw an error, I will remove the Number fn

**Comment by verbanicm** [comment]
@sethvargo Ready for another pass.....

File: src/retry.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { errorMessage } from './errors';
+import { sleep } from './time';
+
+/**
+ * RetryOptions are the config options for the withRetry function
```
Comment (sethvargo) [inline src/retry.ts:21]:
```suggestion
 * RetryOptions are the config options for the withRetry function.
```

File: src/retry.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ */
+export interface RetryOptions {
+  attempts?: number;
```
Comment (sethvargo) [inline src/retry.ts:24]:
```suggestion
  /**
   * attempts is the number of _retry_ attempts. This must be zero or a positive number.
   */
  attempts?: number;
```

File: src/retry.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  const DEFAULT_BACKOFF_MILLISECONDS = 30 * 1000; // 30 seconds
+
+  let attemptLimit = opts?.attempts || DEFAULT_ATTEMPTS;
+  let currentBackoff = opts?.backoff || DEFAULT_BACKOFF_MILLISECONDS;
+  let prevBackoff = 0;
+
+  // recursive execution function
+  const execRetry = async (): Promise<T> => {
+    try {
+      return await fn.call(null);
```
Comment (sethvargo) [inline src/retry.ts:46]: Hmm - this could be unexpected because `this.foo` inside the given function wouldn't resolve anymore.

File: tests/retry.test.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+ */
+
+import 'mocha';
+import { expect } from 'chai';
+
+import { withRetry } from '../src/retry';
+
+describe('retry', () => {
+  describe('#withRetry', () => {
+    const cases = [
```
Comment (sethvargo) [inline tests/retry.test.ts:24]: Can we add some test cases for:

- negative attempts
- zero attempts
- 1 attempt
- negative backoff
- zero backoff

File: tests/time.test.ts [hunk 10]
```diff
@@ -81,4 +81,44 @@ describe('time', () => {
       });
     });
   });
+
+  describe('#sleep', () => {
+    const cases = [
```
Comment (sethvargo) [inline tests/time.test.ts:86]: Can you define the structure explicitly?

```typescript
const cases: {
  name: string;
  input: ...;
}[] = [...]
```

File: tests/retry.test.ts (hunk 9 above)
Comment (sethvargo) [inline tests/retry.test.ts:24]: Can you define the structure explicitly?

**Review by sethvargo — COMMENTED** [review]
Looks good to me. I'm worried about:

1. Some of the edge cases with retries/backoff values
2. Hitting the recursion limit

File: src/retry.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * withRetry a simple implementation of a retry mechanism with fibonacci backoff
+ *
+ * @param fn. A function to retry on failure
+ * @param opts. The retry options
+ * @throws {Error}
+ */
+export async function withRetry<T>(fn: () => Promise<T>, opts?: RetryOptions): Promise<T> {
+  const DEFAULT_ATTEMPTS = 3;
```
Comment (sethvargo) [inline src/retry.ts:49]: Should we move these constants outside the function body so they are only declared on load?

File: src/retry.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  const DEFAULT_BACKOFF_MILLISECONDS = 30 * 1000; // 30 seconds
+  const DEFAULT_MAX_BACKOFF_MILLISECONDS = 60 * 1000 * 5; // 5 minutes
+
+  const backoffLimit = clamp(
```
Comment (sethvargo) [inline src/retry.ts:53]: I'm not sure the `clamp` optimization is worth the complexity it adds here. Are you worried about someone passing in a very large backoff or attempts? We should just document that the function is only well-behaved for reasonable values of attempts and backoffs.

File: src/util.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+/**
+ * clamp ensures a number value is between a min and max value
+ *
+ * @param value. The value to clamp
+ * @param min. The minimum value
+ * @param max. The maxiumum value
+ */
+export function clamp(value: number, min: number, max: number): number {
```
Comment (sethvargo) [inline src/util.ts:24]: I think I'm -1 on this. I'm not sure I see the value of the function or the validation it's used in.

File: src/retry.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+ */
+export interface RetryOptions {
+  /**
+   * retries is the number of _retries_ not the number of attempts. The number
+   * of attempts will be retries + 1. This must be zero or a positive number.
+   */
+  retries: number;
+
+  /**
+   * backoff is the starting backoff time in milliseconds
```
Comment (sethvargo) [inline src/retry.ts:33]:
```suggestion
   * backoff is the starting backoff time, in milliseconds.
```

File: src/retry.ts [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+   */
+  backoff?: number;
+
+  /**
+   * backoffLimit is the maximum backoff time in milliseconds
```
Comment (sethvargo) [inline src/retry.ts:38]:
```suggestion
   * backoffLimit is the maximum backoff time, in milliseconds.
```

File: src/retry.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * withRetry implements a retry mechanism with fibonacci backoff
+ *
+ * @param fn. A function to retry on failure
+ * @param opts. The retry options
+ * @returns fn. A function to start the retry process as a promise
+ * @throws {Error}
+ */
+export function withRetry<T>(fn: () => Promise<T>, opts: RetryOptions): () => Promise<T> {
```
Comment (sethvargo) [inline src/retry.ts:51]: I wonder if we should name this `retry` or `withRetries`. `withRetry` reads weird.

**Review by sethvargo — APPROVED** [review]
Tests are sad - in my experience, you need to add a lot of buffer for OSX for some reason.

File: src/retry.ts (hunk 16 above)
Comment (verbanicm) [reply src/retry.ts:51]: `withRetries` it is!
